### ISSUE TITLE
New rule: `no-implicit-service-injection-argument`

### ISSUE THREAD

**Issue by knownasilya** [ember-cli/eslint-plugin-ember#195]
I'm proposing adding a new Best Practices rule that promotes usage of explicit injections.

So instead of:

```js
store: inject.service()
```

The recommended way would be:

```js
store: inject.service('store')
```

The reasoning behind this change, is that I've heard new Ember developers asking about what that means and where is the file that defines the given injection. It's a sort of "black magic" that allows for shortcuts that don't really help down the line. For example:

```js
users: inject.controller()
// The route also defines `users` as a model.
```

Now if a new developer saw that in the template, even if there wasn't a model defined, they might think it's the model and try to do `{{#each users as |user|}}`.

A better approach:

```js
usersController: inject.controller('users')
// The route defines `users` as a model.
```

The above won't prevent users from naming injections poorly, but it will allow them to be more verbose by default.

This rule could have a fix as well, and would open up the way to deprecating the implicit injections all-together, via an RFC. I'm sure that there are also performance improvements when you no longer have to do that implicit check in the internals.

**Comment by kellyselden** [comment]
I would love to have this rule, but I don't know about adding to an "on by default" group to start.

**Comment by piotrpalek** [comment]
I would personally be against this rule, especially as a default. In my experience people tend to name the same services differently in every component / controller / etc which makes searching for those services tedious.
The opposite way (recommending to use `users: inject.controller()`) would make the code more consistent and in result easier to parse.
It's definitely a tradeoff, but in the spirit of guiding users into the "pit of success" I'd say this approach is a clear winner since long-term the codebase stays more consistent (with the tradeoff being that it can be a bit confusing initially).

**Comment by knownasilya** [comment]
Sure it requires consistency in a way, but its hard to follow in the template or elsewhere that you are working with a service. I think this probably won't be a default rule since most people do it the other way.

The reason I've submitted this is to help with teams that prefer the more explicit method and aren't used to this kind of "magic".

**Comment by knownasilya** [comment]
Give it a test here https://github.com/ember-cli/eslint-plugin-ember/pull/196

**Comment by bmish** [comment]
Related rule: [no-unnecessary-service-injection-argument](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-unnecessary-service-injection-argument.md)

**Comment by rwjblue** [comment]
FWIW, I would love if this rule existed.

**Comment by knownasilya** [comment]
I'll try to get it across the finish line this week.

**Comment by bmish** [comment]
I opened a PR to implement `no-implicit-service-injection-argument`: #1188
